src/server/game/Entities/Player/Player.cpp

```diff
@@ -28832,3 +28835,114 @@ uint8 Player::GetItemLimitCategoryQuantity(ItemLimitCategoryEntry const* limitEn
 
     return limit;
 }
+
+bool Player::IsAtMaxLevel() const
+{
+    return getLevel() >= sWorld->getIntConfig(CONFIG_MAX_PLAYER_LEVEL);
+}
+
+void Player::SetWarModeDesired(bool enabled)
+{
+    // Only allow to toggle on when in stormwind/orgrimmar, and to toggle off in any rested place.
+    // Also disallow when in combat
+    if ((enabled == IsWarModeDesired()) || IsInCombat() || !IsInRestArea())
+        return;
+
+    if (enabled && CanEnableWarModeInArea())
+        return;
+
+    // Don't allow to chang when aura SPELL_PVP_RULES_ENABLED is on
+    if (HasAura(SPELL_PVP_RULES_ENABLED))
+        return;
+
+    if (enabled)
+    {
+        AddPlayerFlag(PLAYER_FLAGS_WAR_MODE_DESIRED);
+        TogglePvpTalents(true);
+        SetPvP(true);
+    }
+    else
+    {
+        RemovePlayerFlag(PLAYER_FLAGS_WAR_MODE_DESIRED);
+        TogglePvpTalents(false);
+        SetPvP(false);
+    }
+
+    UpdateWarModeAuras();
+}
+
+bool Player::IsInFactionFriendlyArea(AreaTableEntry const* inArea /* = nullptr */, AreaTableEntry const** outArea /* = nullptr */) const
```

Comment:
I don't really like this out arg, it would be better to make two separate overloads
```
bool IsInFactionFriendlyArea() const; // this one will just be if (AreaTableEntry const* areaEntry = sAreaTableStore.LookupEntry(GetAreaId()))  IsInFactionFriendlyArea(areaEntry)
bool IsInFactionFriendlyArea(AreaTableEntry const* areaEntry) const;
```